File: py_solutions_11-20/Euler_12.py
# What is the value of the first triangle number to have over five-hundred divisors?

import math
import timeit
import sys


sys.setrecursionlimit(15000)
start = timeit.default_timer()

def factor_gen(tri):
    for i in xrange(2, int(math.sqrt(tri)) + 1):
        if tri % i == 0:
            yield i

def euler_12(n, tri):
    if sum([2 for i in factor_gen(tri)]) + 2 > 500:
        return tri
    return euler_12(n + 1, sum([tri, n, 1]))

print "Answer: %s" % euler_12(1001, sum(xrange(1002)))
stop = timeit.default_timer()
print "Time: %f" % (stop - start)

File: Java/java_solutions_11-20/Euler_12.java
/*
/   Project Euler problem #12 Highly divisible triangular number
/
/   What is the value of the first triangle number to have over five hundred
/   divisors?
/
*/

import static java.lang.Math.sqrt;
import java.util.stream.IntStream;


public class Euler_12
{
    public static long findTriangleDivisors(long triangleNumber) {
        int divisors = 0;
        int triangleLimit = (int) sqrt(triangleNumber);
        IntStream triangleRange = IntStream.range(1, triangleLimit)
                                           .filter(x -> triangleNumber % x == 0);
        return triangleRange.count() * 2;
    }

    public static long findHighlyDivisibleTriangle(int limit, long count, long triangleNumber) {
        if (findTriangleDivisors(triangleNumber) > limit) {
            return triangleNumber;
        }

        return findHighlyDivisibleTriangle(limit, count + 1, triangleNumber + count);
    }

    public static void main(String[] args)
    {
        long start = System.nanoTime();
        long answer = findHighlyDivisibleTriangle(500, 1, 0);
        long stop = System.nanoTime();
        System.out.println("Answer: " + answer);
        System.out.printf("Time: %.4f\n", ((float) stop - start) / 1_000_000_000);
    }
}
